projects/miopen/test/gtest/bn_spatial_test.cpp

```diff
@@ -1,1367 +1,1443 @@
+                                  dshift.desc,
+                                  dshift.desc,
+                                  scale_dev.get(),
+                                  nullptr,
+                                  dscale_dev.get(),
+                                  dshift_dev.get(),
+                                  epsilon,
+                                  savedMean_dev.get(),
+                                  savedInvVar_dev.get(),
+                                  actDesc);
+
+        dx_out.data = handle.Read<T>(dx_out_dev, dx_out.data.size());
+        dscale.data = handle.Read<U>(dscale_dev, dscale.data.size());
+        dshift.data = handle.Read<U>(dshift_dev, dshift.data.size());
+
+        if constexpr(MIO_BN_TIME_EVERYTHING)
+        {
+            auto t_end = std::chrono::high_resolution_clock::now();
+
+            std::cout << "Wall clock: GPU backward_bn_spatial_use_saved pass time: "
+                      << std::chrono::duration<double>(t_end - t_start).count() << " seconds."
+                      << std::endl;
+        }
+
+        return std::make_tuple(dx_out, dscale, dshift);
+    }
+
+    void fail() const
+    {
+        FAIL() << "Backward Batch Spatial Normalization Use Saved Mean and Variance: " << std::endl
+               << "X Input tensor: " << x_input.desc.ToString() << std::endl
+               << "Delta Y Input tensor: " << dy_input.desc.ToString() << std::endl;
+    }
+};
+
+template <typename T>
+constexpr bool
+    is_floating_point_tensor = (std::is_same_v<typename T::value_type, half_float::half> ||
+                                std::is_floating_point_v<typename T::value_type>);
+
+template <class... Ts>
+constexpr bool any_float_tensors()
+{
+    return (is_floating_point_tensor<Ts> || ...);
+}
+
+template <class LeftTuple, class RightTuple, std::size_t... I>
+bool CompareFloatTensorTuples(LeftTuple const& left,
+                              RightTuple const& right,
+                              double tolerance,
+                              std::index_sequence<I...>)
+{
+    return ((miopen::rms_range(std::get<I>(left).data, std::get<I>(right).data) <
+             (std::numeric_limits<
+                  typename std::decay_t<decltype(std::get<I>(left).data)>::value_type>::epsilon() *
+              tolerance)) &&
+            ...);
+}
+
+template <class... CpuT, class... GpuT>
+std::enable_if_t<!(any_float_tensors<CpuT...>() || any_float_tensors<GpuT...>()), bool>
+Compare(std::tuple<CpuT...> const& cpu_result, std::tuple<GpuT...> const& gpu_result, double)
+{
+    return cpu_result == gpu_result;
+}
+
+template <class... CpuT, class... GpuT>
+std::enable_if_t<(any_float_tensors<CpuT...>() || any_float_tensors<GpuT...>()), bool> Compare(
+    std::tuple<CpuT...> const& cpu_result, std::tuple<GpuT...> const& gpu_result, double tolerance)
+{
+    return CompareFloatTensorTuples(
+        cpu_result, gpu_result, tolerance, std::index_sequence_for<CpuT...>{});
+}
```

Comment:
These can be replaced with single function, something like this
```suggestion
template <class... CpuT, class... GpuT>
bool Compare(
    std::tuple<CpuT...> const& cpu_result, std::tuple<GpuT...> const& gpu_result, double tolerance)
{
    if constexpr(any_float_tensors<CpuT...>() || any_float_tensors<GpuT...>())
    {
        return CompareFloatTensorTuples(
            cpu_result, gpu_result, tolerance, std::index_sequence_for<CpuT...>{});
    } else {
        return cpu_result == gpu_result;
    }
}
```

Reply:
Yeah, I need to try it out. I was stuck with sfinae for some reason, so left it in this form. Maybe also, create another header for those, since I'm porting another test, which needs the same functions

Reply:
Yeah, works